### ISSUE TITLE
aggregate_metadata: removes (non)obsolete metadata files

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#4395]
I was trying to re-enable a test which was disabled in the past claiming (in the comments, not commit message) some metadata incompatibility (see d4cf4d8b4e94da8713b9d84f7bc6e8f29e6ee96f), and while at it tried my staple search query, to just get surprised to see pretty much no results.  As you can see below it boils down to the commit in datasets.datalad.org dataset which removed metadata files claiming them being obsolete.

<details>
<summary>Example:</summary> 

Install dataset, and get to the commit which removed "obsolete" files:
```shell
$> datalad install /// && cd datasets.datalad.org && git checkout b5e2df5794c1282daed61de6b1fa64df700a7a5f
install(ok): /tmp/datasets.datalad.org (dataset)                                                                                                 
Note: switching to 'b5e2df5794c1282daed61de6b1fa64df700a7a5f'.

You are in 'detached HEAD' state. You can look around, make experimental
...
HEAD is now at b5e2df5 [DATALAD] Remove obsolete metadata object files

$> datalad search haxby                                                                                   
datalad search haxby  97.12s user 42.97s system 108% cpu 2:09.12 total

$> datalad --version
datalad 0.12.5.dev24
```
</details>

in above - `search` using current maint version (0.12.5.dev24) takes WAY too long (97sec), returns without any output/feedback.  While at it `top` shows datalad appearing once in a while at 10% CPU load, but otherwise seems to be awaiting for some IO...

<details>
<summary>Looking at debug log - it seems spending time assigning those (non existing) metadata object files into the main dataset</summary> 

```shell
[DEBUG  ] Determined class of decorated function: <class 'datalad.local.subdatasets.Subdatasets'> 
[DEBUG  ] Resolved dataset for subdataset reporting/modification: /tmp/datasets.datalad.org 
[DEBUG  ] Query subdatasets of <Dataset path=/tmp/datasets.datalad.org> 
[DEBUG  ] <AnnexRepo path=/tmp/datasets.datalad.org (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done <AnnexRepo path=/tmp/datasets.datalad.org (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Not reporting result (excluded by filter <function get_result_filter.<locals>._fx at 0x7fe93aaa83b0> with arguments {'path': None, 'dataset': <Dataset path=/tmp/datasets.datalad.org>, 'fulfilled': None, 'recursive': True, 'recursion_limit': None, 'contains': PosixPath('/tmp/datasets.datalad.org/.datalad/metadata/objects/31/ds-af94a808f634ec508c87a5162964e2'), 'bottomup': False, 'set_property': None, 'delete_property': None, 'on_failure': 'ignore', 'result_filter': <function is_ok_dataset at 0x7fe93bd18200>} [utils.py:keep_result:610]): {'action': 'subdataset', 'path': '/tmp/datasets.datalad.org/.datalad/metadata/objects/31/ds-af94a808f634ec508c87a5162964e2', 'status': 'impossible', 'message': 'path not contained in any matching subdataset'} 
[DEBUG  ] Determined class of decorated function: <class 'datalad.local.subdatasets.Subdatasets'> 
[DEBUG  ] Resolved dataset for subdataset reporting/modification: /tmp/datasets.datalad.org 
[DEBUG  ] Query subdatasets of <Dataset path=/tmp/datasets.datalad.org> 
[DEBUG  ] <AnnexRepo path=/tmp/datasets.datalad.org (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done <AnnexRepo path=/tmp/datasets.datalad.org (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Not reporting result (excluded by filter <function get_result_filter.<locals>._fx at 0x7fe93aaa84d0> with arguments {'path': None, 'dataset': <Dataset path=/tmp/datasets.datalad.org>, 'fulfilled': None, 'recursive': True, 'recursion_limit': None, 'contains': PosixPath('/tmp/datasets.datalad.org/.datalad/metadata/objects/a6/ds-99480ae1070de795c287271f85aee3'), 'bottomup': False, 'set_property': None, 'delete_property': None, 'on_failure': 'ignore', 'result_filter': <function is_ok_dataset at 0x7fe93bd18200>} [utils.py:keep_result:610]): {'action': 'subdataset', 'path': '/tmp/datasets.datalad.org/.datalad/metadata/objects/a6/ds-99480ae1070de795c287271f85aee3', 'status': 'impossible', 'message': 'path not contained in any matching subdataset'} 
[DEBUG  ] Determined class of decorated function: <class 'datalad.local.subdatasets.Subdatasets'> 
[DEBUG  ] Resolved dataset for subdataset reporting/modification: /tmp/datasets.datalad.org 
[DEBUG  ] Query subdatasets of <Dataset path=/tmp/datasets.datalad.org> 
[DEBUG  ] <AnnexRepo path=/tmp/datasets.datalad.org (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
^C[DEBUG  ] Terminating process for ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] upon exception: KeyboardInterrupt() [selectors.py:select:415] 
ERROR: 
Interrupted by user while doing magic: KeyboardInterrupt() [selectors.py:select:415]

$> ls -ld /tmp/datasets.datalad.org/.datalad/metadata/objects/31/ds-af94a808f634ec508c87a5162964e2
ls: cannot access '/tmp/datasets.datalad.org/.datalad/metadata/objects/31/ds-af94a808f634ec508c87a5162964e2': No such file or directory

$> git show --numstat b5e2df5794c1282daed61de6b1fa64df700a7a5f | grep  ds-af94a808f634ec508c87a5162964e2 
0	168	.datalad/metadata/objects/31/ds-af94a808f634ec508c87a5162964e2
```
</details>

<details>
<summary>If I checkout state before removal of "obsolete" files, things look and behave much better (checkout takes awhile). It doesn't try to figure out if those absent file belong to some subdataset, just `get`s them and returns search result (cold run with `get` - 13 sec, subsequent 2.5 sec).</summary> 

```shell
$> time datalad search Haxby
search(ok): /tmp/datasets.datalad.org/hbnssi (dataset)
search(ok): /tmp/datasets.datalad.org/labs/gobbini/famface/data (dataset)
search(ok): /tmp/datasets.datalad.org/labs/haxby (dataset)
search(ok): /tmp/datasets.datalad.org/labs/haxby/attention (dataset)
search(ok): /tmp/datasets.datalad.org/labs/haxby/life (dataset)
search(ok): /tmp/datasets.datalad.org/labs/haxby/raiders (dataset)
search(ok): /tmp/datasets.datalad.org/openfmri/ds000105 (dataset)
search(ok): /tmp/datasets.datalad.org/openfmri/ds000233 (dataset)
search(ok): /tmp/datasets.datalad.org/openfmri/ds000241 (dataset)
search(ok): /tmp/datasets.datalad.org/openneuro/ds000105 (dataset)
search(ok): /tmp/datasets.datalad.org/openneuro/ds000233 (dataset)
search(ok): /tmp/datasets.datalad.org/openneuro/ds000241 (dataset)
action summary:
  search (ok: 12)
datalad search Haxby  2.58s user 0.48s system 103% cpu 2.961 total
```
</details>

<details>
<summary>Interesting fact is that only metadata files were removed, no changes to `aggregate_v1.json`</summary> 

```shell
$> git show --numstat b5e2df5794c1282daed61de6b1fa64df700a7a5f | grep -v objects
commit b5e2df5794c1282daed61de6b1fa64df700a7a5f
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Thu Mar 12 17:26:28 2020 -0400

    [DATALAD] Remove obsolete metadata object files
```

so - no files listed
</details>

<details>
<summary>and it is only 2 datasets metadata files left:</summary> 

```shell
$> grep "objects/" .datalad/metadata/aggregate_v1.json | sed -e 's,[",],,g' | while read f; do /bin/ls ".datalad/metadata/$f" > /dev/null 2>&1 && echo $f ok; done
objects/da/cn-8d9c7577490ce2d3ef81bccf0c1a7a.xz ok
objects/da/ds-8d9c7577490ce2d3ef81bccf0c1a7a ok
objects/61/cn-48667e6e29c6b0a97dc01db751e641.xz ok
objects/61/ds-48667e6e29c6b0a97dc01db751e641 ok

$> grep "objects/" .datalad/metadata/aggregate_v1.json | sed -e 's,[",],,g' | while read f; do /bin/ls ".datalad/metadata/$f" > /dev/null 2>&1 || echo $f nok; done  | nl | tail
  1339	objects/87/cn-8860472d72b8631fdd13b189d0b314.xz nok
  1340	objects/87/ds-8860472d72b8631fdd13b189d0b314 nok
  1341	objects/f0/cn-3bee1f3d11686ff46b205b8686a4ae.xz nok
  1342	objects/f0/ds-3bee1f3d11686ff46b205b8686a4ae nok
  1343	objects/36/cn-008eea765a7199ca96c3f27244ffe9.xz nok
  1344	objects/36/ds-008eea765a7199ca96c3f27244ffe9 nok
  1345	objects/84/cn-67d4d07a092ef46cdaa22d9b0e7e00.xz nok
  1346	objects/84/ds-67d4d07a092ef46cdaa22d9b0e7e00 nok
  1347	objects/de/cn-e40b46f06f81928c6b9eddae1b5a0e.xz nok
  1348	objects/de/ds-e40b46f06f81928c6b9eddae1b5a0e nok

```
</details>

<details>
<summary>so metadata for everything but one subdataset (and root) was removed, and that dataset is `labs`:
</summary> 

```shell
$> jql .datalad/metadata/aggregate_v1.json| grep -3 48667e6e29c6b0a97dc01db751e641
Failed to detect file type of .datalad/metadata/aggregate_v1.json.
    "refcommit": "313513f66a7d9ab8d3ed6c3e41444237a61c1174"
  },
  "labs": {
    "content_info": "objects/61/cn-48667e6e29c6b0a97dc01db751e641.xz",
    "datalad_version": "0.10.0.rc5.dev18",
    "dataset_info": "objects/61/ds-48667e6e29c6b0a97dc01db751e641",
    "extractors": [
      "datalad_core",
      "annex",
```
</details>

although I do not see any explicit commit adjusting labs around that point
<details>
<summary></summary> 

```shell
607f096 Updated RawDataBIDS which what is updateable (no permission denied or completely removed keys)
927a6ba [DATALAD] Dataset aggregate metadata update
b5e2df5 (HEAD) [DATALAD] Remove obsolete metadata object files
4896a04 adding hcp-openaccess
fdec1ee containers was updated
```
</details>
Since we do not record datalad version used to produce changes (see fresh #4394) -- not yet sure if that could have been due to some older datalad or some new one to blame.

I think altogether there are multiple issues, which are yet to be troubleshooted separately and filed:

- [ ] whatever command lead to this was buggy.  most likely -- `aggregate-metadata`. Likely ran with a specific path. (this one severity-important if not critical, thus marking this issue important for now, later - just move label there)

- [ ] we should report if some (number of them? paths they cover?) metadata objects are not available (at all, or not `get`able) whenever they should be  (UX)

- [ ] avoid lengthy questioning of the files under the path which we know does not have any subdatasets where they potentially belong. (OPT)

**Comment by yarikoptic** [comment]
additional piece I just mentioned:  subsequent to that commit did prune entries within aggregate_v1.json:
```
commit 927a6ba36f0b5ebe9cd0e80d3ba72e94b4024070
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Thu Mar 12 17:29:25 2020 -0400

    [DATALAD] Dataset aggregate metadata update

 .datalad/metadata/aggregate_v1.json                               | 13225 +--------------------------------------------------------------------
 .datalad/metadata/objects/93/cn-b60201b43e1ded020ea31f0b5478ad.xz |     1 +
 .datalad/metadata/objects/93/ds-b60201b43e1ded020ea31f0b5478ad    |     1 +
 labs                                                              |     2 +-
```
and that one was updating labs -- so that is the likely cause indeed, I did re-aggregated only `labs` and it killed all the rest

**Comment by mih** [comment]
Reevaluate 2022-03-01

**Comment by christian-monch** [comment]
@yarikoptic, @mih:

After looking at the issue, and without having the executed command available, I would assume that the faulty metadata processing was done with datalad's metadata code. Given this premise, a found a few issues that coincide, an UX issue and one or two bugs:

1. UX-issue: `datalad aggregate-metadata` was most likely called in `///` with specifying `--update-mode=all`, the path `labs`, and without specifying `-r`, `--recursive`, or `--incremental`. Something like:
```
datalad aggregate-metadata --updated-mode=all labs
```

That would remove metadata from all datasets except the root dataset and the `labs` datasets. The `labs`-metadata contains `haxby` only as subdataset id in the `haspart` relation that is present in the `datalad_core`-generated dataset-level metadata. The file-level metadata object for `labs` is empty. It seems that search did not pick up the `haspart`-relation  (in the version that was used).

Removing the metadata might be considered an UX-issue, because the command line looks quite "innocent".


2. Bug one: there was probably a bug in the version that was used to perform `datalad aggregate-metadata`, something in the range  that prevented the deletion of "dangling" entries in `aggregate_v1.json`

3. Bug two: there was probably a bug in the version that was used to perform `datalad search` because `haxby` should have been picked up from the dataset-level metadata. The cause for missing `haxby` might also be the faulty `aggregate_v1.json`-file.

 
With the current datalad master (`27548d0fa0ae40167d5e9dfb1f9b280020ec774a` 2022-02-23) the bugs are fixed. I get the following results:

```
> git -P log --oneline --no-abbrev -1
4896a0498afafc08286867a1e0bbc8ed3fb160e9 (HEAD -> test-issue) adding hcp-openaccess

> wc -l .datalad/metadata/aggregate_v1.json 
13258 .datalad/metadata/aggregate_v1.json

> find .datalad/metadata/objects|wc -l
1633

> datalad aggregate-metadata --update-mode=all labs
[INFO   ] Attempt to use pre-aggregate metadata for absent /home/cristian/tmp/issue-4395/datasets.datalad.org/labs from dataset at /home/cristian/tmp/issue-4395/datasets.datalad.org 
[INFO   ] Update aggregate metadata in dataset at: /home/cristian/tmp/issue-4395/datasets.datalad.org 
aggregate_metadata(ok): /home/cristian/tmp/issue-4395/datasets.datalad.org (dataset)                                                                                                                                              
[INFO   ] Attempting to save 6 files/datasets 
save(ok): . (dataset)                                                                                                                                                                                                             
action summary:                                                                                                                                                                                                                   
  aggregate_metadata (ok: 1)
  save (ok: 1)

> wc -l .datalad/metadata/aggregate_v1.json 
39 .datalad/metadata/aggregate_v1.json

> find .datalad/metadata/objects|wc -l
64

> datalad search haxby
search(ok): /home/cristian/tmp/issue-4395/datasets.datalad.org/labs (dataset)

> git -P log --oneline --no-abbrev -3
63d118561dcec22953b58c7220eee44739519b02 (HEAD -> test-issue) [DATALAD] Dataset aggregate metadata update
0135b44a957562e9517bc6940a765938234a6a22 [DATALAD] Remove obsolete metadata object files
4896a0498afafc08286867a1e0bbc8ed3fb160e9 adding hcp-openaccess
```

Everything seems to work as expected (if you expect your metadata to be deleted when specifying `--update-mode=all`  ;-) )

In conclusion, I would propose to close this issue.

We might want to open an UX issue for datalad to remove the `--update-mode` option or add a warning and a force-option to `aggregate-metadata`. We might as well remove all metadata-related code from datalad since we have metalad.

**Comment by mih** [comment]
This sound sensible. Let's try closing...
